### ISSUE TITLE
XDG directory functions do not respect XDG variables on macOS

### ISSUE THREAD

**Issue by uncenter** [casey/just#2434]
Due to long standing difficulties with (a/the) maintainer of dirs-rs, they refuse to recognize XDG environment variables explicitly defined by the user on macOS: https://github.com/dirs-dev/directories-rs/issues/47. `config_directory()` will always go to `/Users/<username>/Library/Application Support` on macOS, even if you have set `XDG_CONFIG_HOME`. I would suggest using https://github.com/lunacookies/etcetera instead, though there may be other alternatives that also respect user preferences.

**Comment by laniakea64** [comment]
Changing the behavior of the existing functions would break [`just`'s backwards-compatibility guarantee](https://just.systems/man/en/backwards-compatibility.html).  This would need to be a new function (maybe named `xdg_config_directory()`?)

**Comment by uncenter** [comment]
Yeah that's sound good, we should still update the documentation for the normal `config_directory()` functions to clarify that they return the path for the operating system specification - right now they are literally called "XDG Directories", that name (and description) would be fitting for `xdg_config_directory()` instead :)

**Comment by uncenter** [comment]
Well, dirs is used for other features like global justfile locating:

https://github.com/casey/just/blob/4f31853835afcd8d511fccf7b2b6680f6a628446/src/search.rs#L14-L19

But since that wouldn't affect the backwards compatability of the justfiles themselves would that be a change that could be made at some point as well?

**Comment by casey** [comment]
Regarding #2439, I see the benefit, but it seems unfortunate to add a bunch more functions that are very close to existing functions.

An alternative option would be to add a setting, like `set use-xdg-directories`, which would change the behavior of directory functions to respect the `XDG` variables.

**Comment by soc** [comment]
@uncenter 

> Due to long standing difficulties with (a/the) maintainer of dirs-rs

Please explain.

**Comment by uncenter** [comment]
> @uncenter
> 
> > Due to long standing difficulties with (a/the) maintainer of dirs-rs
> 
> Please explain.

Not sure what you want me to say. Read the issue you just closed and locked conversation on. That kind of response is both childish and a clear example of why moving away from your library is necessary in the first place. There is no reason *not* to respect environment variables that must be EXPLICITLY DEFINED by the user - [it's an intentional disrespect to their choices](https://github.com/dirs-dev/directories-rs/issues/47#issuecomment-596467168). If you want to have this conversation, let's unlock that issue and continue over there as to not bother the maintainers of this repository.

**Comment by soc** [comment]
Look, I do not have to tolerate targeted harassment for maintaining a library in my own free time and ensuring that the library does exactly what it promises on the tin to its users.

You are not the first person who
- thinks their opinion trumps the needs of millions of users of the library that rely on it doing what they chose the library for,
- throws a fit when their demands aren't met,
- goes around filing issues against projects using dirs and
- tries to talk shit behind my back because they cannot respect the choices I have made in my own project.

If you think your behavior is even remotely acceptable, four days before Christmas, please seek appropriate help from a professional, i. e. not me. Because I'm not going to deal with it in any way, shape or form any longer.

> If you want to have this conversation

I really don't. Go away.

**Comment by casey** [comment]
I'm going to hide as off topic all comments related to what `dirs` should or shouldn't do. This is the repo for `just`, and discussion of what other libraries should or shouldn't do is off topic.

I'm actually not sure `just` should make any changes here.

If users want to use the value of the `$XDG_*` variables, they can do `env('XDG_CONFIG_HOME')`, which is short and explicit, and if they don't, they can use the provided functions, like `home_directory()`.

In any case, the documentation should be changed to reflect what the current functions actually do, which is provide paths to system directories.

**Comment by casey** [comment]
Updated the readme in #2535.

**Comment by casey** [comment]
I think since `just` provides an easy way to get environment variables, I'm going to close this for now.

I think it's not a bad situation that the functions always return system-specified locations, and users that want the XDG locations can use `env`.

If we want, we could additionally document the names of the corresponding `XDG_*` environment variables are for each existing function.

**Comment by malikoth** [comment]
I came here because I found that `just` is not finding my global justfile.

```
❯ uname -v
Darwin Kernel Version 21.6.0: Mon Jun 24 00:56:10 PDT 2024; root:xnu-8020.240.18.709.2~1/RELEASE_X86_64

❯ echo $XDG_CONFIG_HOME
/Users/malikoth/.local/etc

❯ la ~/.local/etc/just
total 4.0K
-rw-r--r-- 1 malikoth staff 2.3K Mar 28 21:30 justfile

❯ just -g
error: Global justfile not found
```

As observed above, not only does this affect the recipes in the justfile (which I can understand opting not to change), but also `just`'s ability to find its global justfile on macOS when the user has explicitly set a value for $XDG_CONFIG_HOME, all of which apply to me. 

(Why is .local used for most home directory stuff, except for config files? Why stuff those into ~/.config? That's never made sense to me! .local to me represents the user-scoped equivalent of all the directories in a Unix filesystem root, including etc, so I move it all into .local/etc. I have a shell init script explicitly setting all XDG_* env vars, and another script dedicated exclusively to setting env vars to "help" the apps that need a little more assistance dealing with that fact, but it's not too bad.)

So for me, following the documentation precisely, just does not behave as described in the documentation. I think this is a bug and should be reopened, and solved.
